src/main/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverter.java

```diff
@@ -15,56 +15,729 @@
+
+		if (customTarget.isPresent()) {
+			return conversionService.convert(value, customTarget.get());
+		}
+
+		return Enum.class.isAssignableFrom(value.getClass()) ? ((Enum<?>) value).name() : value;
+	}
+
+	@SuppressWarnings("unchecked")
+	protected Object getWriteComplexValue(ElasticsearchPersistentProperty property, TypeInformation<?> typeHint,
+			Object value) {
+
+		if (typeHint.isCollectionLike() || value instanceof Iterable) {
+			return writeCollectionValue(value, property, typeHint);
+		}
+		if (typeHint.isMap()) {
+			return writeMapValue((Map<String, Object>) value, property, typeHint);
+		}
+
+		if (property.isEntity() || !isSimpleType(value)) {
+			return writeEntity(value, property, typeHint);
+		}
+
+		return value;
+	}
+
+	private Object writeEntity(Object value, ElasticsearchPersistentProperty property, TypeInformation<?> typeHint) {
+
+		Document target = Document.create();
+		writeEntity(mappingContext.getRequiredPersistentEntity(value.getClass()), value, target,
+				property.getTypeInformation());
+		return target;
+	}
+
+	private Object writeMapValue(Map<String, Object> value, ElasticsearchPersistentProperty property,
+			TypeInformation<?> typeHint) {
+
+		Map<Object, Object> target = new LinkedHashMap<>();
+		Streamable<Entry<String, Object>> mapSource = Streamable.of(value.entrySet());
+
+		if (!typeHint.getActualType().getType().equals(Object.class)
+				&& isSimpleType(typeHint.getMapValueType().getType())) {
+			mapSource.forEach(it -> target.put(it.getKey(), getWriteSimpleValue(it.getValue())));
+		} else {
+
+			mapSource.forEach(it -> {
+
+				Object converted = null;
+				if (it.getValue() != null) {
+
+					if (isSimpleType(it.getValue())) {
+						converted = getWriteSimpleValue(it.getValue());
+					} else {
+						converted = getWriteComplexValue(property, ClassTypeInformation.from(it.getValue().getClass()),
+								it.getValue());
+					}
+				}
+
+				target.put(it.getKey(), converted);
+			});
+		}
+
+		return target;
+	}
+
+	private Object writeCollectionValue(Object value, ElasticsearchPersistentProperty property,
+			TypeInformation<?> typeHint) {
+
+		Streamable<?> collectionSource = value instanceof Iterable ? Streamable.of((Iterable<?>) value)
+				: Streamable.of(ObjectUtils.toObjectArray(value));
+
+		List<Object> target = new ArrayList<>();
+		if (!typeHint.getActualType().getType().equals(Object.class) && isSimpleType(typeHint.getActualType().getType())) {
+			collectionSource.map(this::getWriteSimpleValue).forEach(target::add);
+		} else {
+
+			collectionSource.map(it -> {
+
+				if (it == null) {
+					return null;
+				}
+
+				if (isSimpleType(it)) {
+					return getWriteSimpleValue(it);
+				}
+
+				return getWriteComplexValue(property, ClassTypeInformation.from(it.getClass()), it);
+			}).forEach(target::add);
+
+		}
+		return target;
+	}
+
+	@Override
+	public <T> T mapToObject(String source, Class<T> clazz) throws IOException {
```

Comment:
Can we get rid of that method to eliminate `ObjectMapper` or should we move that task into a separate ticket?

Reply:
This will be part of the cleanup when all the Mapper stuff moved here